text/0004-structured_data_io.md

```diff
@@ -12,33 +11,49 @@ we'd hope, or less complete than we'd expect from a new RFC.
 # Summary
 [summary]: #summary
 
-Structured I/O defines interfaces for reading & writing streams of _entries_,
-which are elements of a composite data structure such as an array or object.
-Structured reader & writer interfaces combine a byte stream, data format 
-and schema to create entry readers & writers that produce & consume entries
-instead of bytes. Structured I/O streams can be composed & connected to form
-the basis of rich data communication capable of spanning across formats.
+Structured I/O defines interfaces for reading & writing streams of parsed 
+ _entries_, which are elements of dynamic-yet-structred data such as JSON, CBOR,
+or CSV documents. Structured reader & writer interfaces combine a byte stream, 
+data format and schema to create entry readers & writers that produce & consume 
+entries of parsed primtive types instead of bytes. Structured I/O streams can be 
+composed & connected to form the basis of rich data communication capable of 
+spanning across formats.
 
 # Motivation
 [motivation]: #motivation
+
+One of the prime goals of qri is to to be able to make any dataset comparable to
+another dataset. Datasets are also intended to be a generic-yet-structured
+document format, able to support all sorts of data with varying degrees of
+quality. These requirements mean datasets must be able to define their own 
+schemas, and may include data that violates that schema.
```

Comment:
this may be a dumb question but, what does a schema tell me about the data if it's the case that it may or may not be invalid? (seeing as we have access to previous schemas, that were valid).  In a typical world where a schema can't be invalid, I interpret the schema as the source of truth.  If a change invalidates it, and let's say removes the schema or rebuilds it from the data dynamically, I can both go back to the old version to see what the truth used to be and look at the current (possibly absent) schema and know what the truth is now.  But if the schema may or may not be invalid I'm not sure what I'm supposed to do with that.

Reply:
No a dumb question at all!

In the case where data may or may not be valid, the schema is a pointer to the _authorial intent_ of the dataset. It's the equivalent of "this is how the data _should_ look". In this context, the schema isn't a source of truth, it's a source of _intent_ that is only a source of _truth_ if all data matches, or we're operating in "strict mode".  

That "should look like" can be very useful in different context for higher-level decision making. For example, the schema may dictate that a field labeled "date" should be a datestamp in the format `YYYY-MM-DD`. I can use the schema to make smarter decisions about how to coerce invalid data. At this level of API design, however, "smarter decisions" should be kicked up the chain for different implementations to worry about, because one size doesn't fit all, and it'd be too tough to know how to solve those specific problems in generalized ways. So instead we focus on an interface that makes sure the schema is on hand should you want to use it in a higher-level API. We don't require you to make use of it at all if it doesn't help a given use case.

Finally, it's not always true that we have access to previous schemas that were valid.  Peers don't always have all of a dataset history on hand. Even when we _do_ have the full history on hand, there is no guarantee that a completely valid schema ever existed in the dataset history, which is why we fall back to identity schemas instead of getting entangled with previous maybe-or-maybe-not-fully-valid schemas, because one of the two Identity schemas is _required_ to be valid to create a dataset in the first place.